### ISSUE TITLE
.NET MAUI does not work with Android 13 Android.permission.write_external_storage

### ISSUE THREAD

**Issue by FatCodeMonkey** [dotnet/maui#11275]
### Description

My App was working and able to take pictures on Android 12 and .NET 6 but as soon as I upgraded to Android 13 and .NET 7, it stopped working. When I call this method:

FileResult photo = await MediaPicker.Default.CapturePhotoAsync();

I get this error:  Microsoft.Maui.ApplicationModel.PermissionException: 'StorageWrite permission was not granted: Denied'.

I do have these declarations in the AndroidManifest.xml file:

	<uses-permission android:name="android.permission.CAMERA" />
	<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
	<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
	<queries>
		<intent>
			<action android:name="android.media.action.IMAGE_CAPTURE" />
		</intent>
	</queries>

### Steps to Reproduce

1. Create a new .NET MAUI App with the default template (not empty template).
2. Replace the code in the OnCounterClicked method with the code below.
3. Connect your device to your PC (my device is Google Pixel Pro 6).
4. Run (Debug) the app on the local Android 13 device.
5. Click the Counter button
6. Notice the error

### Link to public reproduction project repository

https://github.com/FatCodeMonkey/MAUIPhoto

### Version with bug

6.0.486 (current)

### Last version that worked well

6.0.424

### Affected platforms

Android

### Affected platform versions

Android 13

### Did you find any workaround?

No

### Relevant log output

```shell
0xFFFFFFFFFFFFFFFF in Android.Runtime.JNIEnv.monodroid_debugger_unhandled_exception	C#
 	0x1A in Android.Runtime.JNINativeWrapper._unhandled_exception at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:12,5	C#
 	0x1D in Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:23,26	C#
 	0x17 in System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw	C#
 	0x6 in System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0	C#
 	0xC in Android.App.SyncContext. at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.App/SyncContext.cs:36,19	C#
 	0xE in Java.Lang.Thread.RunnableImplementor.Run at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Thread.cs:36,6	C#
 	0x8 in Java.Lang.IRunnableInvoker.n_Run at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net6.0/android-33/mcw/Java.Lang.IRunnable.cs:84,4	C#
 	0x8 in Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:22,5	C#
```

**Comment by UkeHa** [comment]
@FatCodeMonkey does it work if you add android:requestLegacyExternalStorage="true" to your manifest?
<application android:allowBackup="true" android:icon="@mipmap/appicon" android:roundIcon="@mipmap/appicon_round" android:supportsRtl="true" **android:requestLegacyExternalStorage="true"**>></application>

**Comment by Stedy59** [comment]
What baffles me, is that .net MAUI was released with an initial target Android API of 33. But the underlying code does not support Android API 33!

Starting from Android 13, if your application target SDK is specified to 33 or above, the READ_EXTERNAL_STORAGE permission will be completely useless, and applying for it will have no effect.

Correspondingly, Google has added three runtime permissions: READ_MEDIA_IMAGES, READ_MEDIA_VIDEO, and READ_MEDIA_AUDIO, which are used to manage the photos, videos and audio files of the phone respectively. That is to say, in the past, you only needed to apply for a READ_EXTERNAL_STORAGE permission. This is no longer possible. You have to apply for it on demand, so that users can learn more precisely which media permissions your app has applied for.

<manifest ...>
    
    
    <uses-permission android:name="android.permission.READ_MEDIA_IMAGES" />
    <uses-permission android:name="android.permission.READ_MEDIA_AUDIO" />
    <uses-permission android:name="android.permission.READ_MEDIA_VIDEO" />

    
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"
                     android:maxSdkVersion="32" />
    <application ...>
        ...
    </application>
</manifest>

It is the same for notifications!
<manifest ...>
    <uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
    <application ...>
        ...
    </application>
</manifest>

I have raised this issue multiple times, but one would think that since they are targeting API 33 from the start, that this would be a priority!

I pulled the Android permissions code, and updated it to support API 33. Hopefully they will realize that making the "Essentials" an actual part of the MAUI core, requires a higher level of attention to external API changes.

**Comment by sidewinder94** [comment]
@Stedy59 Could you share your modifications while we wait for the MAUI team to fix this ? (right now as a workaround I fixed the target android sdk to 30 in the android manifest).

**Comment by YMichurin** [comment]
I do not see how I can circumvent the issue either. Everyone around me has devices with the latest Android installed which I can't run my .net maui app on. The app can't work without media picker (MediaPicker control), but it seems to rely on WRITE_EXTERNAL_STORAGE only, which is not in the list of android permissions anymore. It does not care about new api 33 permissions like READ_MEDIA_IMAGES, it wants WRITE_EXTERNAL_STORAGE and does not run camera without it. Am I stuck till the issue is fixed .net maui or there is some temp. workaround for that?

**Comment by sidewinder94** [comment]
@YMichurin Unless you need what api 33 need, che'ge the target SDK inside the android manifest to target api level 32. That did it for me

**Comment by YMichurin** [comment]
hi @sidewinder94, if I do change the manifest to api 32, would my app work on a phone with api 33 installed?

**Comment by Stedy59** [comment]
Here is a snippet that I use to open the device gallery and/or select a ringtone...

`[assembly: UsesPermission(Name = "android.permission.READ_EXTERNAL_STORAGE", MaxSdkVersion = 32)]
[assembly: UsesPermission(Name = "android.permission.READ_MEDIA_AUDIO")]
[assembly: UsesPermission(Name = "android.permission.READ_MEDIA_IMAGES")]
[assembly: UsesPermission(Name = "android.permission.READ_MEDIA_VIDEO")]
namespace StedySoft.Shared {

	#region Class MediaServices
	public static partial class MediaServices {

		#region Declarations
		internal const int OPENGALLERY = 100;
		internal const int OPENRINGTONES = 101;
		#endregion

		#region Private Methods
		private static bool PlatformGetExternalStoragePermissions(string requestMessage, int permissionType = OPENGALLERY) {
			try {
				Permission status = Permission.Denied;
				if (Build.VERSION.SdkInt >= BuildVersionCodes.Tiramisu) {
					switch (permissionType) {
						case OPENGALLERY:
							status = ContextCompat.CheckSelfPermission(Application.Context, Manifest.Permission.ReadMediaImages);
							break;
						case OPENRINGTONES:
							status = ContextCompat.CheckSelfPermission(Application.Context, Manifest.Permission.ReadMediaAudio);
							break;
					}
				}
				else {
					status = ContextCompat.CheckSelfPermission(Application.Context, Manifest.Permission.ReadExternalStorage);
				}
				if (status != Permission.Granted) {
					if (Build.VERSION.SdkInt >= BuildVersionCodes.Tiramisu) {
						switch (permissionType) {
							case OPENGALLERY:
								if (ActivityCompat.ShouldShowRequestPermissionRationale(Platform.Activity, Manifest.Permission.ReadMediaImages)) { Toast.Show(requestMessage, ToastDuration.Long); }
								ActivityCompat.RequestPermissions(Platform.Activity, new[] { Manifest.Permission.ReadMediaImages }, 0);
								break;
							case OPENRINGTONES:
								if (ActivityCompat.ShouldShowRequestPermissionRationale(Platform.Activity, Manifest.Permission.ReadMediaAudio)) { Toast.Show(requestMessage, ToastDuration.Long); }
								ActivityCompat.RequestPermissions(Platform.Activity, new[] { Manifest.Permission.ReadMediaAudio }, 0);
								break;
						}
					}
					else {
						if (ActivityCompat.ShouldShowRequestPermissionRationale(Platform.Activity, Manifest.Permission.ReadExternalStorage)) { Toast.Show(requestMessage, ToastDuration.Long); }
						ActivityCompat.RequestPermissions(Platform.Activity, new[] { Manifest.Permission.ReadExternalStorage }, 0);
					}
					Stopwatch stopwatch = new();
					stopwatch.Start();
					do {
						Thread.Sleep(10);
						if (Build.VERSION.SdkInt >= BuildVersionCodes.Tiramisu) {
							switch (permissionType) {
								case OPENGALLERY:
									status = ContextCompat.CheckSelfPermission(Application.Context, Manifest.Permission.ReadMediaImages);
									break;
								case OPENRINGTONES:
									status = ContextCompat.CheckSelfPermission(Application.Context, Manifest.Permission.ReadMediaAudio);
									break;
							}
						}
						else {
							status = ContextCompat.CheckSelfPermission(Application.Context, Manifest.Permission.ReadExternalStorage);
						}
						if (stopwatch.ElapsedMilliseconds > 10000) {
							throw new Exception("Timed out waiting for permission.");
						}
					} while (status != Permission.Granted);
				}
				return true;
			}
			catch (Exception ex) {
				Logger.LogException("MediaServices", "PlatformGetExternalStoragePermissions", ex);
			}
			return false;
		}
`

**Comment by sidewinder94** [comment]
@YMichurin Yes, no issues there, since devices running higher versions know how to deal with older ones.

**Comment by espenrl** [comment]
And here is the code
```
<uses-sdk android:minSdkVersion="21" android:targetSdkVersion="32" />
```

**Comment by FM1973** [comment]
Same here.
I had to set the targetSdkVersion to 32 to use Notifications.
A lot of people already user Android 13. It would be nice to have the new permissions implemented in Maui.

**Comment by HopInTheCloud** [comment]
I have the same issue. The question now is, how long will it take for them to release an update to fix this. Releasing updates seems to be taking a really long time.

**Comment by espenrl** [comment]
I avoided this issue by using Xamarin.MediaGallery instead.

https://github.com/dimonovdd/Xamarin.MediaGallery

**Comment by symbiogenesis** [comment]
> I avoided this issue by using Xamarin.MediaGallery instead.
> 
> https://github.com/dimonovdd/Xamarin.MediaGallery

I tried this, and followed the readme file, with both the stable and preview release, but it just pops up a dialog saying "Can't connect to the camera" in the Android 13 Emulator with SDK level 33.

Are you on .NET6 like the sample? Is there some other trick to it?

**Comment by UkeHa** [comment]
@symbiogenesis, again, if you don't **need** to target sdk33 just target 32 and it'll work just fine. Still a bug that needs to be fixed.

**Comment by symbiogenesis** [comment]
Ok, this seems to be working now.

> @symbiogenesis, again, if you don't **need** to target sdk33 just target 32 and it'll work just fine. Still a bug that needs to be fixed.

Thank you. I had tried this before and it didn't work for some reason, but I tried it now and it works. This is the best solution.

**Comment by davispalla** [comment]
I'm having the same problem. I'm following the colleague's last suggestion, that is, changing the destination to sdk 32.
Previously, I had used Xam.Media.Plugin, however, when I open the camera, vs closes and the app in debug mode also closes without throwing the error message.

**Comment by Ghostbird** [comment]
> @FatCodeMonkey does it work if you add android:requestLegacyExternalStorage="true" to your manifest? <application android:allowBackup="true" android:icon="@mipmap/appicon" android:roundIcon="@mipmap/appicon_round" android:supportsRtl="true" **android:requestLegacyExternalStorage="true"**>>

After you update your app to target Android 11, the system ignores the requestLegacyExternalStorage flag. [source](https://developer.android.com/about/versions/11/privacy/storage)

So no, that's a completely different option that only worked before Android 11.

**Comment by Ghostbird** [comment]
> And here is the code
> 
> ```
> <uses-sdk android:minSdkVersion="21" android:targetSdkVersion="32" />
> ```

How do you get that to work? If I do that I get:

```
warning XA4211: AndroidManifest.xml //uses-sdk/@android:targetSdkVersion '32' is less than $(TargetFrameworkVersion) ''. Using API-33 for ACW compilation. 
```

The result is still built using API 33 and it doesn't work.

I tried adding a `TargetFrameworkVersion` option with value 32 to the build, but that didn't change a thing.

**Comment by Ghostbird** [comment]
> > > And here is the code
> > > ```
> > > <uses-sdk android:minSdkVersion="21" android:targetSdkVersion="32" />
> > > ```
> > 
> > 
> > How do you get that to work? If I do that I get:
> 
> Double check that Targeting Android option is on in the project settings and it targets lower version. This is what I have in the project file and it works. I do not think you need anything else behind that to avoid the error
> 
> ```
> <SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">21.0</SupportedOSPlatformVersion>
> ```

Hmm… yes, that option is already in our csproj file, it's needed to build Android regardless of current API version.

**Comment by symbiogenesis** [comment]
I just ignored the XA4211 warning. It worked. I took photos successfully on Android 13, both with the real device and the emulator.

I think your PR is still very helpful, and I hope you continue to clean up that horrible code.

The iOS code is possibly even worse. The photos don't respect the device orientation, and it is using obsolete APIs with lots of TODOs and pragmas in the code.

**Comment by Ghostbird** [comment]
> I just ignored the XA4211 warning. It worked. I took photos successfully on Android 13, both with the real device and the emulator.

I tried, but the message is on point. It's still built using API 33 and therefore the error still occurs.

**Comment by symbiogenesis** [comment]
I still have that warning in my code right now, and it is the only one. Thinking again, I think it is possible that I only verified it using the emulator, and took the device photos with an ipad.

I recently had tried switching to the Xamarin.MediaGallery nuget package (with MAUI support) because of a different ios camera bug regarding device orientation data not getting saved in the EXIF metadata.

That completely didn't work for me at all, so I tried switching back to MAUI essentials plus a workaround, and that seems to have fixed ios. But the Android version isn't even launching now. So I cannot really verify anything about this right now, except to say that it is working perfect in the emulator for Android SDK 33

![image](https://user-images.githubusercontent.com/1724472/213638425-42e806e6-5c92-4b36-beed-f5af1443db64.png)
.

**Comment by symbiogenesis** [comment]
Actually, I'm now seeing 3 warnings in the error list

![image](https://user-images.githubusercontent.com/1724472/213639289-78d21add-51ce-42d7-b89d-1dd0732ec547.png)

**Comment by Ghostbird** [comment]
I do see those too, note that the third error seems to be a re-parse of the second one, where the `Code` is parsed separately from the `Description` by Visual Studio. I'm just seeing the two unique warnings.

**Comment by Ghostbird** [comment]
Ok, I managed to find it, and it has to do with a really unexpected twist in how the android manifest interprets certain values.

I followed this thread, an applied appropriate changes to the manifest, such as the relevant parts of [this comment](#issuecomment-1312023057) and [this comment](https://github.com/dotnet/maui/issues/11275#issuecomment-1336197989).

Therefore my android manifest ended up with these two lines (other lines omitted):

```xml
  <uses-sdk android:minSdkVersion="21"
    android:targetSdkVersion="32" />
  <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" 
    android:maxSdkVersion="32" />
```

The really obscure thing is that you can set `android:targetSdkVersion="32"`, get the warning that it compiles with API 33 after all. Apparently this means that the `uses-permission` is omitted, because we're at API 33. If I remove the `android:maxSdkVersion` attribute from the `uses-permission` line, I still build API 33 according to the warning, but now suddenly `uses-permission` _does_ work correctly, even though that permission doesn't even exist in API 33.

I'm baffled, yet happy that our release-blocker can be bypassed this way.

**Comment by ejboustany** [comment]
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android">
	<application android:allowBackup="true" android:icon="@mipmap/appicon" android:roundIcon="@mipmap/appicon_round" android:supportsRtl="true"></application>
	<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
	<uses-permission android:name="android.permission.INTERNET" />
	<uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
	<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
	<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />



	<uses-permission android:name="android.permission.READ_MEDIA_IMAGES" />
	<uses-permission android:name="android.permission.READ_MEDIA_AUDIO" />
	<uses-permission android:name="android.permission.READ_MEDIA_VIDEO" />
	<uses-permission android:name="android.permission.CAMERA" />

	<queries>
		<intent>
			<action android:name="android.media.action.IMAGE_CAPTURE" />
		</intent>
	</queries>
</manifest>

This is how my AndroidManifest.xml looks like and its still not working any ideas?

**Comment by Ghostbird** [comment]
> ```
> <uses-permission android:name="android.permission.READ_MEDIA_IMAGES" />
> <uses-permission android:name="android.permission.READ_MEDIA_AUDIO" />
> <uses-permission android:name="android.permission.READ_MEDIA_VIDEO" />
> <uses-permission android:name="android.permission.CAMERA" />
> 
> <queries>
> 	<intent>
> 		<action android:name="android.media.action.IMAGE_CAPTURE" />
> 	</intent>
> </queries>
> ```
> 
> This is how my AndroidManifest.xml looks like and its still not working any ideas?

You haven't yet told us what you're trying to do. What exactly is not working? Are you trying to take a picture on Android 13? Then reread this thread, because that's discussed quite extensively.
You can remove the `READ_MEDIA` permissions, since taking pictures is not about reading, but writing. Then you'll need:

```xml
  <uses-sdk android:minSdkVersion="21"
    android:targetSdkVersion="32" />
  <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"  />
```

Although you may use a different `minSdkVersion`.

**Comment by ejboustany** [comment]
> 

I have added that before it didn't work for some reason. When I removed the READ_MEDIA permissions its not working. Thanks for the help. Yes I was trying to capture images on Android 13. That was my first time asking a question on GitHub and looks like ill be doing it more now. Happy to join the MAUI community.

**Comment by AHComp** [comment]
```
        public async Task SaveMediaAsync(byte[] data, string name, string location = "EsiMedias")
        {
            var documentsPath = Environment.GetExternalStoragePublicDirectory(Environment.DirectoryPictures)?.Path;
            if (documentsPath != null)
            {
                documentsPath = System.IO.Path.Combine(documentsPath, location);
                Directory.CreateDirectory(documentsPath);

                var filePath = System.IO.Path.Combine(documentsPath, name);

                await File.WriteAllBytesAsync(filePath, data);
            }
        }
```
If i try to override a file it was not working and i got a access denied message !

I used the workaround with api 32 in manifest.

So creating of the file was working but not override !

**Comment by joseluisct** [comment]
Same problem here, using the workaround and now it's working with Android 13 device/emulator. 
Waiting for the update to target again android 13.
Thank you for the workaround!

**Comment by guillaumelimoo** [comment]
Hi there, I'm glad to hear that the issue has been resolved in .NET 8 Preview 1. However, does this mean that the fix will not be included in a service release before the stable version of .NET 8 is released?
Thanks for the information 🙏.

**Comment by Ghostbird** [comment]
If you follow the links where this issue is referenced, you _can_ find [this answer to your question](https://github.com/dotnet/maui/issues/13119#issuecomment-1421349485) but it's not very clear.

In summary, it just _barely_ missed the previous .NET 7 service release, and it will be in the upcoming one.

**Comment by guillaumelimoo** [comment]
Very clear, thank you! Hopefully soon 🤞

**Comment by samirgcofficial** [comment]
Waiting for the Solution Still.

**Comment by Ghostbird** [comment]
Please read the thread before commenting. It contains a workaround that will tide you over, and you'll see that the fix has been incorporated into MAUI .NET 8 and .NET 7 and will be in the next release.

**Comment by HurseyNZ** [comment]
> Please read the thread before commenting. It contains a workaround that will tide you over, and you'll see that the fix has been incorporated into MAUI .NET 8 and .NET 7 and will be in the next release.

Not to discredit what you say here, but don't know if you realise that the work around mentioned in this thread might actually not be applicable to everyone.  I for one don't get the option to simply change the target framework without changing the target .net runtime which then breaks dependencies.

**Comment by Ghostbird** [comment]
I don't get why you would have to change the target framework. The workaround applies to the `MediaPicker.CaptureAsync` bug, and involves a downgrade of the Android SDK, from API 33 which is supported in .NET 7.0+ to SDK 32 which is supported in .NET 6.0+.

My thought are:

- If you're on .NET 6.0 I don't think you can be on API 33 which has this problem, but if you do, you can safely downgrade to API 32.
- If you're on .NET 7.0 you can safely downgrade to API 32.
- Which leaves the possibility that you're forced to be on a .NET 8.0 pre-release. If that's the case, I can't help you, but given that you're on a pre-release version you can expect some bugs. The .NET 8.0 release will not have the bug.

Please explain where I went wrong, and I might be able to help you find a working solution.

**Comment by guillaumelimoo** [comment]
Unfortunately, some individuals, including myself, are unable to revert from Android version 33 to version 32 due to the utilization of specific libraries that require the former and its accompanying new permissions, particularly ZXing. I am wondering if there is a guarantee that we will be able to access these new permissions in a Service Release for .NET 7. 
Your message was edited, and you acknowledged some uncertainty regarding this matter. Thank you for your assistance.

**Comment by Ghostbird** [comment]
Interesting, the app I'm building has no trouble with ZXing.Net.Maui and ZXing.Net.Maui.Controls, both version 0.3.0-preview.1 on Android API 32.

There are no new permissions in the service release. In API 33 some permissions have been removed. The bug I'm referring is that the `MediaPicker.CaptureAsync` method required one of those permissions. This broke API 33 compatibility as Android will always return _permission denied_ as that permission no longer exists.

**Comment by HurseyNZ** [comment]
> * If you're on .NET 7.0 you can safely downgrade to API 32 until the next service release.

This is my issue, I'm on .net7 but unable to select any target api other than 33, the options for other api levels simply are not there via the project properties.  
![image](https://user-images.githubusercontent.com/20807483/225112537-09610c3f-1f70-4fa3-bb03-6cc120007b3b.png)

I'm quite happy to admit I could be overlooking something, perhaps you'd be kind enough to put a guide together on how to do this?

**Comment by sidewinder94** [comment]
@HurseyNZ You need to manually edit the android manifest file and edit the "android:targetSdkVersion" property to have a value <= 32.

You WILL have a warning in your build stating that you only have the SDK for API level 33. This is expected and should have no consequences

**Comment by guillaumelimoo** [comment]
> Interesting, the app I'm building has no trouble with ZXing.Net.Maui and ZXing.Net.Maui.Controls, both version 0.3.0-preview.1 on Android API 32.

Can you explain me the procedure to follow knowing that the package explicitly targets Android 33?

<img width="339" alt="Capture d’écran 2023-03-15 à 10 17 58" src="https://user-images.githubusercontent.com/1955115/225263885-432c3673-bc7a-4139-a4f4-05fafea268b3.png">

**Comment by AHComp** [comment]
Hey does anybody try to publish a app downgraded to api 32 ?
I got an error on try to publish !

![image](https://user-images.githubusercontent.com/35998144/225289720-7153d80c-3c8c-47d0-9dbd-3ff515f236de.png)

**Comment by Ghostbird** [comment]
@HurseyNZ I'm not familiar with the tooling you're using. I'm using plain `dotnet` to build. It may be that your IDE is preventing you from taking actions that are non-standard, even if completely valid. As @guillaumelimoo says, you can change the setting in `AndroidManifest.xml`.

@AHComp The warnings about `XA1006`, `XA4211` are valid, and do not block publishing. These warnings are shown because the situation is undesirable. It is however not _problematic_ and in this case done with purpose as a temporary workaround.
From what I'm seeing, some other warnings may be obscured because of a missing translation to German. Ich weiß das die Deutschen Deutsch lieben. However, in this case you may want to try to find a way to run in English, just to verify whether there's an error that cannot be shown in German.

**Comment by Ghostbird** [comment]
@guillaumelimoo Is that targeting actually causing problems? Because it doesn't cause me any problems. Note that as the error says, ACW compilation still uses API-33. I'm not sure what ACW compilation is, apparently to an extent it still uses API 33. In my case I avoid the issues caused by the permissions, yet ZXing still works properly. The relevant section of my `AndroidManifest.xml` was shown earlier in the thread. I had some issues myself getting the configuration right. In my case I had to ensure there was no `maxSdkVersion` in `<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"  />`. I thought this was counter-intuitive, but it I think it is because of that ACW compilation thing.

**Comment by Ghostbird** [comment]
No, this is a MAUI problem, not a .NET problem. We're waiting for the next .NET _MAUI_ 7.0 service release.

**Comment by AHComp** [comment]
Ok and the date is ?

**Comment by Ghostbird** [comment]
Unknown. It's running late. Maybe because of the upcoming .NET MAUI 8
release.

**Comment by HurseyNZ** [comment]
> @HurseyNZ I'm not familiar with the tooling you're using. I'm using plain `dotnet` to build. It may be that your IDE is preventing you from taking actions that are non-standard, even if completely valid. As @sidewinder94 says, you can change the setting in `AndroidManifest.xml`.

It's Visual Studio 2022 I'm using.  Getting to the exact same point as @AHComp.  Have tried many different versions of target SDK in the app manifest and project code, all result in Using API-33 for ACW Compliance

**Comment by PruntschSoftware** [comment]
I am not debugging on a emulator but on a real device that uses Android 13.0 - API 33. Would it still help to downgrade the target SDK or am i doomed because the phone has API 33 ?

**Comment by Ghostbird** [comment]
It works fine. We're using this in our production app.

**Comment by janseris** [comment]
> Hey does anybody try to publish a app downgraded to api 32 ? I got an error on try to publish !
> 
> ![image](https://user-images.githubusercontent.com/35998144/225289720-7153d80c-3c8c-47d0-9dbd-3ff515f236de.png)

Google Play has now policy that before August 31 2023, all new apps must target API 33.  
That's ~3 months before .NET 8 will be released.
For this, releasing this in .NET 8 is late.
@samhouts I can see that you added label "fixed in .NET 7.0". What service release will this be available in?

**Comment by AHComp** [comment]
@Redth How long it could take to add the service release for .Net 7 to visual studio ?

**Comment by Ghostbird** [comment]
Also it is separate from the .NET service release. Unless the MAUI service
release is dependent on fixes in .NET. Then a .NET service release must
first be released, before a MAUI service release can be made. I don't
expect that to be the case.

**Comment by Ghostbird** [comment]
My point being that you _may_ do that. But MAUI is a tool completely
separate from Visual Studio and does not _need_ to wait on anything related
to Visual Studio.

In fact I suspect that MAUI isn't even installed by Visual Studio, but by
the Visual Studio installer, which is another tool altogether.

I have been using MAUI for several months now, yet I haven't touched Visual
Studio in years.

Using Visual Studio for MAUI is like having a garage attached to your home.
MAUI is the car. Sure it is convenient to park your car in the garage.
However, you don't need a garage to park your car. You also in general
don't need to renovate your garage if you buy a new car. They're two
separate things, that are sometimes nice to use together.

**Comment by FM1973** [comment]
@Ghostbird 
How do you install the latest version of maui without a visual studio update?
As I´ve read, latest Visual Studio Versions contain a lot of bugs. But I´d like to have the lates maui release.

**Comment by Ghostbird** [comment]
The easiest way in my opinion is using the .NET CLI. Personally I prefer
this as it works in almost all environments, also in CI environments such
as Github Actions. The .NET CLI is normally installed with the .NET SDK
which you probably have installed already.

See this [`dotnet workload install` manual](
https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-workload-install)

**Comment by FM1973** [comment]
@Ghostbird 
So you say to install the latest .net core SDK (7.0.4) and then use dotnet workload update. Am I right?

**Comment by guillaumelimoo** [comment]
Congratulations on the new release pack and thanks!

**Comment by thatkidyousaw1** [comment]
Anyone knows why this is still not working? Here's my dotnet SDK and MAUI versions:

![image](https://user-images.githubusercontent.com/35909220/235593826-6b108e1e-d4fa-4e6f-8cc6-27b6158776ca.png)
![image](https://user-images.githubusercontent.com/35909220/235593862-61865893-ec79-4ee9-ba78-7cc82a49a204.png)

**Comment by Ghostbird** [comment]
> Anyone knows why this is still not working? Here's my dotnet SDK and MAUI versions:
> 
> ![image](https://user-images.githubusercontent.com/35909220/235593826-6b108e1e-d4fa-4e6f-8cc6-27b6158776ca.png) ![image](https://user-images.githubusercontent.com/35909220/235593862-61865893-ec79-4ee9-ba78-7cc82a49a204.png)

Your .NET SDK version is actually newer than mine, since you're running the preview. I think you're running a Visual Studio specific version of the dotnet workloads. I don't have experience with those.

You could try to install the maui workloads using the `dotnet` CLI.

In my case it lists this:
```bash
$ dotnet workload list
dotnet workload list
Installed Workload Id      Manifest Version      Installation Source
--------------------------------------------------------------------
maui-android               7.0.81/7.0.100        SDK 7.0.200
```

If `dotnet` is already in your `PATH` you can run. `dotnet workload install maui-android`. Replace `maui-android` with the appropriate other workloads to install those.

**Comment by thatkidyousaw1** [comment]
> > Anyone knows why this is still not working? Here's my dotnet SDK and MAUI versions:
> > ![image](https://user-images.githubusercontent.com/35909220/235593826-6b108e1e-d4fa-4e6f-8cc6-27b6158776ca.png) ![image](https://user-images.githubusercontent.com/35909220/235593862-61865893-ec79-4ee9-ba78-7cc82a49a204.png)
> 
> Your .NET SDK version is actually newer than mine (7.0.203), since you're running the preview. I think you're running a Visual Studio specific version of the dotnet workloads. I don't have experience with those.
> 
> You could try to install the maui workloads using the `dotnet` CLI.
> 
> In my case it lists this:
> 
> ```shell
> $ dotnet workload list
> Installed Workload Id      Manifest Version      Installation Source
> --------------------------------------------------------------------
> maui-android               7.0.81/7.0.100        SDK 7.0.200
> ```
> 
> If `dotnet` is already in your `PATH` you can run. `dotnet workload install maui-android`. Replace `maui-android` with the appropriate other workloads to install those.

I managed to install it now using dotnet CLI, however, the same error occurs. I updated the AndroidManifest file too. Did you manage to make it work on your end? :o

![image](https://user-images.githubusercontent.com/35909220/235604509-724bc05c-cd7d-4c60-83d0-6ef15e1c1104.png)

**Comment by Ghostbird** [comment]
Yeah, our latest production release uses Android 13. I see that you're still on the .NET preview. Why are you trying this using a preview? I'd use the latest stable if I were you. _If_ it works using the stable version, but not using the preview, please report a regression.

Did you follow the instructions higher up in this thread? If so, your Android manifest should contain these two lines (although your `minSdkVersion` may be different:

```xml
<uses-sdk android:minSdkVersion="21"
  android:targetSdkVersion="33" />
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" android:maxSdkVersion="32"/>
```

If you've tried all that and it still doesn't help, please post the actual error.

**Comment by thatkidyousaw1** [comment]
> Yeah, our latest production release uses Android 13. I see that you're still on the .NET preview. Why are you trying this using a preview? I'd use the latest stable if I were you. _If_ it works using the stable version, but not using the preview, please report a regression.
> 
> Did you follow the instructions higher up in this thread? If so, your Android manifest should contain these two lines (although your `minSdkVersion` may be different):
> 
> ```
> <uses-sdk android:minSdkVersion="21"
>   android:targetSdkVersion="33" />
> <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" android:maxSdkVersion="32"/>
> ```
> 
> If you've tried all that and it still doesn't help, please post the actual error.

Thank you so much for the support, it's actually for the FolderPicker function. This is the error:

[DOTNET] Folder is not picked, You need to declare using the permission: `android.permission.READ_EXTERNAL_STORAGE` in your AndroidManifest.xml

This is my manifest:
![image](https://user-images.githubusercontent.com/35909220/235611427-4c8c230e-9484-4d30-9e21-17cd8d2d87b2.png)
Thank you!

**Comment by zZHorizonZz** [comment]
[Workaround](https://github.com/CommunityToolkit/Maui/issues/998#issuecomment-1531142942)

**Comment by Ghostbird** [comment]
This issue is about `android.permission.WRITE_EXTERNAL_STORAGE`. As I think @zZHorizonZz implies, it seems your problem is with `android.permissions.READ_EXTERNAL_STORAGE`. In that case, this is not the right issue for you, but the workaround mentioned above may work.

**Comment by zZHorizonZz** [comment]
> This issue is about `android.permission.WRITE_EXTERNAL_STORAGE`. As I think @zZHorizonZz implies, it seems your problem is with `android.permissions.READ_EXTERNAL_STORAGE`. In that case, this is not the right issue for you, but the workaround linked above may work.

Yeah, but since he asked about that in my issue as well, I don't really know if he has a problem with write or read permissions. Anyway, I tested it and it worked for both folder selection and file writing.
